File: scripts/governance/code_health_check.py
#!/usr/bin/env python3
"""Code Health — orphan file detection for Python source"""

import ast
import sys
from pathlib import Path


def get_source_files(directory: Path) -> list[Path]:
    return [
        f
        for f in directory.rglob("*.py")
        if not any(p in str(f) for p in ["__pycache__", ".venv", "venv", "__init__"])
        and "test_" not in f.name
    ]


def get_all_imports(files: list[Path]) -> set[str]:
    imported = set()
    for filepath in files:
        try:
            tree = ast.parse(filepath.read_text())
        except SyntaxError:
            continue
        for node in ast.walk(tree):
            if isinstance(node, ast.ImportFrom) and node.module:
                parts = node.module.split(".")
                candidate = Path(*parts)
                imported.add(str(candidate) + ".py")
                imported.add(str(candidate / "__init__.py"))
    return imported


app_dir = Path("app")
if not app_dir.exists():
    print("✓ Code health: no app/ directory")
    sys.exit(0)

source_files = get_source_files(app_dir)
if len(source_files) <= 1:
    print("✓ Code health: single-file project — skip orphan check")
    sys.exit(0)

imported = get_all_imports(source_files)
orphans = [
    f
    for f in source_files
    if not any(imp_path in str(f) for imp_path in imported)
    and f.name not in ("config.py", "settings.py", "main.py", "app.py")
]

if orphans:
    print(f"⚠ Code health: {len(orphans)} potentially orphaned file(s):")
    for f in orphans:
        print(f"  {f}")

print(f"✓ Code health: {len(source_files)} source files analyzed")

File: scripts/governance/code-health-check.js
#!/usr/bin/env node
/**
 * Code Health — orphan file + dead export detection.
 * Warns (non-blocking) about files never imported and exports never used.
 */

const fs = require('fs');
const path = require('path');

function getSourceFiles(dir) {
  const files = [];
  if (!fs.existsSync(dir)) return files;
  function walk(d) {
    for (const f of fs.readdirSync(d)) {
      const full = path.join(d, f);
      const stat = fs.statSync(full);
      if (stat.isDirectory() && !['node_modules', 'dist', '.git', 'coverage', '__tests__'].includes(f)) {
        walk(full);
      } else if (/\.(ts|tsx)$/.test(f) && !f.endsWith('.d.ts') && !f.endsWith('.test.ts')) {
        files.push(full);
      }
    }
  }
  walk(dir);
  return files;
}

function getAllImports(files) {
  const imported = new Set();
  for (const file of files) {
    const content = fs.readFileSync(file, 'utf8');
    const re = /from\s+['"]([^'"]+)['"]/g;
    let m;
    while ((m = re.exec(content)) !== null) {
      if (m[1].startsWith('.')) {
        const resolved = path.resolve(path.dirname(file), m[1]);
        imported.add(resolved);
        imported.add(resolved + '.ts');
      }
    }
  }
  return imported;
}

const sourceFiles = getSourceFiles('src');
if (sourceFiles.length <= 1) {
  console.log('✓ Code health: single-file project — skip orphan check');
  process.exit(0);
}

const imported = getAllImports(sourceFiles);
const orphans = sourceFiles.filter(f => !imported.has(f) && !f.includes('index.ts') && !f.includes('env.ts'));

if (orphans.length > 0) {
  console.warn(`⚠ Code health: ${orphans.length} potentially orphaned file(s):`);
  orphans.forEach(f => console.warn(`  ${path.relative(process.cwd(), f)}`));
  // Non-blocking warning
}

console.log(`✓ Code health: ${sourceFiles.length} source files analyzed`);
process.exit(0);
